crates/containerd-shim-wasmtime/src/http_proxy.rs

```diff
@@ -0,0 +1,183 @@
+// Heavily inspired by wasmtime serve command:
+// https://github.com/bytecodealliance/wasmtime/blob/main/src/commands/serve.rs
+
+use std::net::{IpAddr, SocketAddr};
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+
+use anyhow::{bail, Result};
+use containerd_shim_wasm::container::RuntimeContext;
+use wasmtime::component::ResourceTable;
+use wasmtime::Store;
+use wasmtime_wasi_http::bindings::http::types::Scheme;
+use wasmtime_wasi_http::bindings::ProxyPre;
+use wasmtime_wasi_http::body::HyperOutgoingBody;
+use wasmtime_wasi_http::io::TokioIo;
+use wasmtime_wasi_http::{WasiHttpCtx, WasiHttpView};
+
+use crate::instance::{envs_from_ctx, WasiPreview2HttpCtx};
+
+const DEFAULT_ADDR: SocketAddr =
+    SocketAddr::new(IpAddr::V4(std::net::Ipv4Addr::new(0, 0, 0, 0)), 8080);
+
+const DEFAULT_BACKLOG: u32 = 100;
+
+type Request = hyper::Request<hyper::body::Incoming>;
+
+pub(crate) async fn serve_conn(
+    ctx: &impl RuntimeContext,
+    instance: ProxyPre<WasiPreview2HttpCtx>,
+) -> Result<()> {
+    use hyper::server::conn::http1;
+    let env = envs_from_ctx(ctx);
+
+    let addr = env
+        .iter()
+        .find(|(key, _)| key == "WASMTIME_HTTP_PROXY_SOCKET_ADDR")
+        .and_then(|(_, val)| val.parse().ok())
+        .unwrap_or(DEFAULT_ADDR);
+
+    let backlog = env
+        .iter()
+        .find(|(key, _)| key == "WASMTIME_HTTP_PROXY_BACKLOG")
+        .and_then(|(_, val)| val.parse().ok())
+        .unwrap_or(DEFAULT_BACKLOG);
+
+    let socket = match addr {
+        SocketAddr::V4(_) => tokio::net::TcpSocket::new_v4()?,
+        SocketAddr::V6(_) => tokio::net::TcpSocket::new_v6()?,
+    };
+
+    // Conditionally enable `SO_REUSEADDR` depending on the current
+    // platform. On Unix we want this to be able to rebind an address in
+    // the `TIME_WAIT` state which can happen then a server is killed with
+    // active TCP connections and then restarted. On Windows though if
+    // `SO_REUSEADDR` is specified then it enables multiple applications to
+    // bind the port at the same time which is not something we want. Hence
+    // this is conditionally set based on the platform (and deviates from
+    // Tokio's default from always-on).
+    socket.set_reuseaddr(!cfg!(windows))?;
+    socket.bind(addr)?;
+
+    let listener = socket.listen(backlog)?;
+
+    log::info!("Serving HTTP on http://{}/", listener.local_addr()?);
+
+    let handler = ProxyHandler::new(instance, &env);
+
+    loop {
+        let (stream, _) = listener.accept().await?;
+        log::debug!("New connection");
+
+        let stream = TokioIo::new(stream);
+        let h = handler.clone();
```

Comment:
Yes, I think so. Note that the handler is just an `Arc` wrapper so cloning is cheap (TM).

- first clone is creating a new reference for each incoming connection, otherwise we would move handler and it would be unavailable for the next loop iteration,
- second clone is preventing moving a handler into closure which would prevent server function from handling multiple request. 

I think it helps to think of it as nested loops. This is also consistent with `wasmtime_wasi_http` example that can be found here: https://docs.rs/wasmtime-wasi-http/latest/wasmtime_wasi_http/#examples